scripts/db-heal-test.py

```diff
@@ -0,0 +1,298 @@
+parser.add_argument("--tmp", type=str, help="Path to tmp dir.")
+parser.add_argument("--iter", type=int, help="Number of iterations.", default=1000)
+parser.add_argument("--window", type=int,
+                    help="Delay between start of sync process and forced termination (in seconds).", default=5)
+parser.add_argument("--cp-granularity", type=int,
+                    help="How often will Carmen create checkpoints (in blocks).", default=10)
+parser.add_argument("--make", action=argparse.BooleanOptionalAction,
+                    help="If enabled, necessary Aida tool will be build before the script begins.")
+
+args = parser.parse_args()
+
+aida_path = args.aida
+aida_db_path = args.aida_db
+tmp_path = args.tmp
+number_of_iterations = args.iter
+window = args.window
+checkpoint_granularity = args.cp_granularity
+make = args.make
+
+# Find Carmen path within Aida directory
+# (we always want to use same Carmen version for block processing and recovering)
+carmen_path = os.path.join(aida_path, 'carmen', 'go')
+
+# Mark first checkpoint
+latest_checkpoint = checkpoint_granularity
+
+if not aida_path or aida_path == "":
+    print("please set Aida using --aida")
+    sys.exit(1)
+if not aida_db_path or aida_db_path == "":
+    print("please set AidaDB using --aida-db")
+    sys.exit(1)
+if not tmp_path or tmp_path == "":
+    tmp_path = tempfile.gettempdir()
+    print(f"tmp not set - using default {tmp_path}")
+
+# Block variables
+first_block = 0
+last_block = 60000000
+
+
+# --- Script --- #
+
+# Create working dir which gets deleted after the run
+working_dir = os.path.join(tmp_path, 'db-heal-test')
+if os.path.exists(working_dir):
+    shutil.rmtree(working_dir)
+os.makedirs(working_dir)
+
+# Log file path from which we read output to find kill_block
+aida_log_file = os.path.join(working_dir, 'aida.log')
+carmen_log_file = os.path.join(working_dir, 'carmen.log')
+genesis = os.path.join(working_dir, 'test_genesis.dat')
+
+print("Your settings:")
+print(f"\tNumber of iterations: {number_of_iterations}.")
+print(f"\tSync time before kill: {window} seconds.")
+print(f"\tCheckpoint granularity: {checkpoint_granularity} blocks.")
+
+
+# Function which checks programs return code, if program failed, log is printed and True is returned.
+def has_program_failed(code, log):
+    if code != 0:
+        log.close()
+        with open(carmen_log_file, 'r') as l:
+            text = l.read()
+            print(text)
+        return True
+    return False
+
+
+# Function which checks every line added to aida_log_file and behaves accordingly to the line.
+def check_aida_log(sleep_time: int, checkpoint: int):
+    start = 0.0
+    with open(aida_log_file, 'r') as f:
+        while True:
+            line = f.readline()
+            if not line:
+                time.sleep(0.1)
+                continue
+            if start > 0 and time.time() - start >= sleep_time:
+                print("Interrupting...")
+                process.terminate()
```

Comment:
Python's dynamic scoping makes it difficult to identify the `process` that is actually terminated here => it should be a parameter.